pkg/rbac/cacheValidation.go

```diff
@@ -0,0 +1,230 @@
+		onAdd:    c.clearUserData,
+		onModify: nil, // FIXME: Skipping MODIFY because we are receiving too many events.
+		onDelete: c.clearUserData,
+	}
+	go watchClusterRoleBindings.start(ctx)
+
+	// Watch GROUPS
+	watchGroups := watchResource{
+		dynamicClient: c.shared.dynamicClient,
+		gvr:           schema.GroupVersionResource{Resource: "groups", Group: "user.openshift.io", Version: "v1"},
+		onAdd:         c.clearUserData,
+		onModify:      c.clearUserData,
+		onDelete:      c.clearUserData,
+	}
+	go watchGroups.start(ctx)
+
+	// Watch CRDS
+	watchCRDs := watchResource{
+		dynamicClient: c.shared.dynamicClient,
+		gvr: schema.GroupVersionResource{Resource: "customresourcedefinitions",
+			Group: "apiextensions.k8s.io", Version: "v1"},
+		onAdd:    c.clearUserData,
+		onModify: c.clearUserData,
+		onDelete: nil, // Deletions can wait for normal expiration.
+	}
+	go watchCRDs.start(ctx)
+}
+
+// Start watching for changes to a resource and trigger the action to update the cache.
+func (w watchResource) start(ctx context.Context) {
+	for {
+		watch, watchError := w.dynamicClient.Resource(w.gvr).Watch(ctx, metav1.ListOptions{})
+		if watchError != nil {
+			klog.Warningf("Error watching resource %s. Error: %s", w.gvr.String(), watchError)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+
+		defer watch.Stop()
+		klog.V(2).Infof("Watching resource: %s", w.gvr.String())
+
+		for {
+			select {
+			case <-ctx.Done():
+				klog.V(2).Info("Stopped watching resource. ", w.gvr.String())
+				watch.Stop()
+				return
+
+			case event := <-watch.ResultChan(): // Read events from the watch channel.
+				klog.V(6).Infof("Event: %s \tResource: %s  ", event.Type, w.gvr.String())
+				o, error := runtime.UnstructuredConverter.ToUnstructured(runtime.DefaultUnstructuredConverter, &event.Object)
+				if error != nil {
+					klog.Warningf("Error converting %s event.Object to unstructured.Unstructured. Error: %s",
+						w.gvr.Resource, error)
+				}
+				obj := &unstructured.Unstructured{Object: o}
+
+				switch event.Type {
+				case "ADDED":
+					if w.onAdd != nil {
+						w.onAdd(obj)
+					}
+				case "MODIFIED":
+					if w.onModify != nil {
+						w.onModify(obj)
+					}
+				case "DELETED":
+					if w.onDelete != nil {
+						w.onDelete(obj)
+					}
+
+				default:
+					klog.V(2).Infof("Received unexpected event. Waiting 5 seconds and restarting watch for %s", w.gvr.String())
+					watch.Stop()
+					time.Sleep(5 * time.Second)
+				}
+			}
+		}
+	}
+}
+
+func (c *Cache) clearUserData(obj *unstructured.Unstructured) {
+	if c.pendingInvalidation {
+		klog.V(5).Info("There's a pending request to clear the User Data cache.")
+		return
+	}
+	c.pendingInvalidation = true
+	klog.Info("Invalidating UserData cache. Waiting 5 seconds to 'debounce' or avoid triggering too many requests.")
+
+	go func() {
+		time.Sleep(5 * time.Second)
+
+		c.usersLock.Lock()
+		defer c.usersLock.Unlock()
+		for _, userCache := range c.users {
```

Comment:
@jlpadilla Do we want to invalidate all the users cache when there is on ROLE change for User X ?